server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1623,14 +1625,17 @@ private static boolean canMatchAfterRewrite(final ShardSearchRequest request, fi
     @SuppressWarnings("unchecked")
     public static boolean queryStillMatchesAfterRewrite(ShardSearchRequest request, QueryRewriteContext context) throws IOException {
         Rewriteable.rewrite(request.getRewriteable(), context, false);
-        final boolean aliasFilterCanMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
-        final boolean canMatch;
+        boolean canMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
         if (canRewriteToMatchNone(request.source())) {
-            QueryBuilder queryBuilder = request.source().query();
-            canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
-        } else {
-            // null query means match_all
-            canMatch = aliasFilterCanMatch;
+            if (request.source().query() != null) {
```

Comment:
I looked into this briefly, and I agree this would be a positive change, but I think it's complex enough that I would prefer to do it as a follow up PR.

Reply:
After looking into it a bit more, I think this is quite difficult in a subtle way where we already make certain assumptions if queries isn't empty. That's not to say it can't be done, just we replace query with something like queries is size of 1 and there's no knn searches, etc.

Reply:
This seems odd, I'd like to understand more around why we need that type of conditional. It's counter intuitive to me that if you have multiple queries, you need to do things substantially differently compared to when you have a single one.

Reply:
One other important note that I forgot to mention is we actually use `query` for an aggregation as part of rank. This is a really nice way to convey this information.

The most challenging part of this change is how to handle SearchContext and where I think you would end up doing checks to see if you only have 1 query or not. Maybe it is only one check, though for single query. You'll also need additional checks anywhere you process queries to see if it's 1 or not, but we already have checks anyway for if queries is empty, so maybe I overestimated how hard this would be.

If it's all right I would still prefer to try to make this change as a separate PR.

Reply:
This has been updated to only use `queries` now.

Reply:
thanks a lot for the effort you put into this change, it makes things a whole lot better!